x-pack/plugin/vector-tile/src/main/java/org/elasticsearch/xpack/vectortile/rest/GridAggregation.java

```diff
@@ -0,0 +1,241 @@
+        // 12 16777216 6 14117882 0.8414913416
+        // 13 67108864 6 14117882 0.2103728354
+        // 14 268435456 7 98825162 0.3681524172
+        // 15 1073741824 8 691776122 0.644266719
+        // 16 4294967296 8 691776122 0.1610666797
+        // 17 17179869184 9 4842432842 0.2818666889
+        // 18 68719476736 10 33897029882 0.4932667053
+        // 19 274877906944 11 237279209162 0.8632167343
+        // 20 1099511627776 11 237279209162 0.2158041836
+        // 21 4398046511104 12 1660954464122 0.3776573213
+        // 22 17592186044416 13 11626681248842 0.6609003122
+        // 23 70368744177664 13 11626681248842 0.165225078
+        // 24 281474976710656 14 81386768741882 0.2891438866
+        // 25 1125899906842620 15 569707381193162 0.5060018015
+        // 26 4503599627370500 15 569707381193162 0.1265004504
+        // 27 18014398509482000 15 569707381193162 0.03162511259
+        // 28 72057594037927900 15 569707381193162 0.007906278149
+        // 29 288230376151712000 15 569707381193162 0.001976569537
+        private static final int[] ZOOM2RESOLUTION = new int[] {
+            0,
+            0,
+            0,
+            1,
+            1,
+            2,
+            2,
+            3,
+            3,
+            4,
+            4,
+            5,
+            6,
+            6,
+            7,
+            8,
+            8,
+            9,
+            10,
+            11,
+            11,
+            12,
+            13,
+            13,
+            14,
+            15,
+            15,
+            15,
+            15,
+            15 };
+
+        @Override
+        public GeoGridAggregationBuilder newAgg(String aggName) {
+            return new GeoHexGridAggregationBuilder(aggName);
+        }
+
+        @Override
+        public Rectangle bufferTile(Rectangle tile, int z, int gridPrecision) {
+            if (z == 0 || gridPrecision == 0) {
+                // no need to buffer at level 0 as we are looking to all data.
+                return tile;
+            }
+            final int aggPrecision = gridPrecisionToAggPrecision(z, gridPrecision);
+            if (aggPrecision < 2) {
+                // we need to consider all data
+                return new Rectangle(-180, 180, GeoTileUtils.LATITUDE_MASK, -GeoTileUtils.LATITUDE_MASK);
+            }
+            return new Rectangle(
+                GeoUtils.normalizeLon(tile.getMinX() - LON_BUFFER_SIZE[aggPrecision]),
+                GeoUtils.normalizeLon(tile.getMaxX() + LON_BUFFER_SIZE[aggPrecision]),
+                Math.min(GeoTileUtils.LATITUDE_MASK, tile.getMaxY() + LAT_BUFFER_SIZE[aggPrecision]),
+                Math.max(-GeoTileUtils.LATITUDE_MASK, tile.getMinY() - LAT_BUFFER_SIZE[aggPrecision])
+            );
+        }
+
+        @Override
+        public int gridPrecisionToAggPrecision(int z, int gridPrecision) {
+            return ZOOM2RESOLUTION[GEOTILE.gridPrecisionToAggPrecision(z, gridPrecision)];
```

Comment:
I worry that we are trying to be too clever here, and trying to force the zoom levels of the H3 tiles to match the zoom levels of the grid. What about pushing this problem to the client? So when doing geotiles, the possible zoom values would be 1 to 29, but when doing H3 the possible values are reduced to 0 to 15. The server side code is much simpler, and clients like kibana need to worry about just what works best for visualization?